[JIT] Functional Graph Pass

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33297 [JIT] Allow mutated values as functional graph inputs
* #33199 [JIT] remove list appends
* #33186 [JIT] Pass To Safely Remove Aten Inplace Ops
* **#33020 [JIT] Functional Graph Pass**

This is a pass to create functional blocks. The other PRs in the stack help avoid some of the limitations that are are often found in graphs. It's possible that this would work well with a graph that is frozen. Follow up work items that will help this pass:

- We don't currently have any capacity in alias analysis to tell whether a Value that came from the wildcard set "re-escapes" back into the wildcard set.
- More comments on the semantics of the graph and correctness conditions
- We could consider using dynamic dag if the perf of this is a limitation. 
- potential make Functional Graphs Functional Blocks instead, so that we do not repeatedly copy constants, also to make IR read easier.

Differential Revision: [D20603188](https://our.internmc.facebook.com/intern/diff/D20603188)

Commit:
initial commit

[ghstack-poisoned]

Commit:
Update on "[JIT] [WIP] Functional Graph Pass"


This is a WIP pass to create functional blocks. There are still some limitations in alias analysis that make this pass functionally useless on normal graphs. However, it's possible that this could work well with a graph that is frozen. Follow up work items that will help this pass:

- We don't currently have any capacity in alias analysis to tell whether a Value that came from the wildcard set "re-escapes" back into the wildcard set. This is needed to create larger functional graphs.
- A functionalizing pass that removes ops like "aten::add_" when it preserves semantics, and rewrites list and dict mutation into functional creation when possible
- More comments on the semantics of the graph and correctness conditions
- We could consider using dynamic dag if the perf of this is a limitation. 



[ghstack-poisoned]

suo commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -319,6 +319,7 @@ void AliasDb::analyzeImpl(Node* node) {
       return analyzeLoop(node);
     case prim::FusionGroup:
     case prim::DifferentiableGraph:
+    case prim::FunctionalGraph:
```
do we need to analyze the entire subgraph? I think we can just treat it as a purely functional node.

eellison replied:
I'll do this in a follow up when i add the implicit copies.

suo commented on torch/csrc/jit/passes/create_functional_graphs.h:
```diff
@@ -0,0 +1,14 @@
+namespace torch {
+namespace jit {
+
+TORCH_API void CreateFunctionalGraphs(const std::shared_ptr<Graph>& graph);
```
nit: might as well take a Graph& if you're taking the shared_ptr by const ref

suo commented on torch/csrc/jit/passes/create_functional_graphs.cpp:
```diff
@@ -0,0 +1,233 @@
+
+void CreateFunctionalGraphs(const std::shared_ptr<Graph>& graph) {
+  // Run Constant Pooling so constants get hoisted
+  ConstantPooling(graph);
```
We can leave this in, but as a general principle I want to start enforcing that graph passes don't run preprocessing and can ingest any sort of graph. It's not super important in this case, but I feel like it'll force us to make fewer assumptions which makes the passes more composable.

suo commented on torch/csrc/jit/passes/create_functional_graphs.h:
```diff
@@ -0,0 +1,14 @@
+
+namespace torch {
+namespace jit {
+
```
Can you write a quick high-level description of the pass as a comment?

suo commented on torch/csrc/jit/passes/create_functional_graphs.cpp:
```diff
@@ -0,0 +1,233 @@
+    return changed;
+  }
+
+  bool AnalyzeFunctionalSubset(Node* n) {
```
nit: I found this name to be confusing, because "Analyze" is not well-defined (and doesn't obviously map to a bool-returning function). Maybe `isFunctional`?

suo commented on torch/csrc/jit/passes/create_functional_graphs.cpp:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  bool AnalyzeFunctionalSubset(Node* n) {
+    bool functional_node = true;
```
nit: would prefer if bool started with `is` or similar

suo commented:
Sorry I am not done reviewing, the thing submitted my comments by accident :/

suo commented on torch/csrc/jit/passes/create_functional_graphs.cpp:
```diff
@@ -0,0 +1,233 @@
+      functional_node = functional_node && functional_block;
+    }
+
+    auto inputs = n->inputs();
```
It wasn't immediately obvious to me that `mutated_values_` will be populated with the outputs of previous nodes which are the inputs to this one. Worth a comment maybe.

suo commented on torch/csrc/jit/passes/create_functional_graphs.cpp:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  std::unordered_set<Node*> functional_nodes_;
+  std::unordered_set<Value*> mutated_values_;
```
This seems redundant, since it's basically a cache for a `hasWriters` check. Do things get a lot slower if we remove it? If we must have it, it feels like something better placed in `AliasDb`.

eellison replied:
i had to do this because you're mutating the graph during the pass, so you'll hit assertion errors if you remove a node's input and then ask if it has writers. will comemtn

eellison replied:
There's a comment already actually:
```
    // First, analyze the functional subset of the graph, and then create
    // functional graphs. The graph gets mutated when we create functional
    // subgraphs, invalidating the AliasDb, so we need to do our analysis
    // first.
```

suo reviewed (approved):
Looks like a good start. Left some comments in-line but I don't think any of them are blocking. We definitely should come up a more extensive set of tests to help convince ourselves that this is correct

suo commented on torch/csrc/jit/passes/create_functional_graphs.cpp:
```diff
@@ -0,0 +1,233 @@
+
+  bool AnalyzeFunctionalSubset(Node* n) {
+    bool functional_node = true;
+
```
nodes with `hasSideEffects` should trivially return false I think

eellison replied:
I would like to clarify this in a follow up, because it disallows `prim::Bailout`. not sure if that's what we want

Commit:
Update on "[JIT] Functional Graph Pass"


This is a pass to create functional blocks. The other PRs in the stack help avoid some of the limitations that are are often found in graphs. It's possible that this would work well with a graph that is frozen. Follow up work items that will help this pass:

- We don't currently have any capacity in alias analysis to tell whether a Value that came from the wildcard set "re-escapes" back into the wildcard set.
- More comments on the semantics of the graph and correctness conditions
- We could consider using dynamic dag if the perf of this is a limitation. 
- potential make Functional Graphs Functional Blocks instead, so that we do not repeatedly copy constants, also to make IR read easier.


[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/create_functional_graphs.cpp:
```diff
@@ -0,0 +1,233 @@
+      bool functional = AnalyzeFunctionalSubset(n);
+      is_functional_block = is_functional_block && functional;
+    }
+    return is_functional_block;
```
Yes, correct, there's no way to have a block output that is not also a node output, however a block can have an input that is not an output (e.g. `prim::If`)

Commit:
Update on "[JIT] Functional Graph Pass"


This is a pass to create functional blocks. The other PRs in the stack help avoid some of the limitations that are are often found in graphs. It's possible that this would work well with a graph that is frozen. Follow up work items that will help this pass:

- We don't currently have any capacity in alias analysis to tell whether a Value that came from the wildcard set "re-escapes" back into the wildcard set.
- More comments on the semantics of the graph and correctness conditions
- We could consider using dynamic dag if the perf of this is a limitation. 
- potential make Functional Graphs Functional Blocks instead, so that we do not repeatedly copy constants, also to make IR read easier.

Commit:
Update on "[JIT] Functional Graph Pass"


This is a pass to create functional blocks. The other PRs in the stack help avoid some of the limitations that are are often found in graphs. It's possible that this would work well with a graph that is frozen. Follow up work items that will help this pass:

- We don't currently have any capacity in alias analysis to tell whether a Value that came from the wildcard set "re-escapes" back into the wildcard set.
- More comments on the semantics of the graph and correctness conditions
- We could consider using dynamic dag if the perf of this is a limitation. 
- potential make Functional Graphs Functional Blocks instead, so that we do not repeatedly copy constants, also to make IR read easier.

Commit:
Update on "[JIT] Functional Graph Pass"


This is a pass to create functional blocks. The other PRs in the stack help avoid some of the limitations that are are often found in graphs. It's possible that this would work well with a graph that is frozen. Follow up work items that will help this pass:

- We don't currently have any capacity in alias analysis to tell whether a Value that came from the wildcard set "re-escapes" back into the wildcard set.
- More comments on the semantics of the graph and correctness conditions
- We could consider using dynamic dag if the perf of this is a limitation. 
- potential make Functional Graphs Functional Blocks instead, so that we do not repeatedly copy constants, also to make IR read easier.

Commit:
Update on "[JIT] Functional Graph Pass"


This is a pass to create functional blocks. The other PRs in the stack help avoid some of the limitations that are are often found in graphs. It's possible that this would work well with a graph that is frozen. Follow up work items that will help this pass:

- We don't currently have any capacity in alias analysis to tell whether a Value that came from the wildcard set "re-escapes" back into the wildcard set.
- More comments on the semantics of the graph and correctness conditions
- We could consider using dynamic dag if the perf of this is a limitation. 
- potential make Functional Graphs Functional Blocks instead, so that we do not repeatedly copy constants, also to make IR read easier.

Differential Revision: [D20603188](https://our.internmc.facebook.com/intern/diff/D20603188)
